### ISSUE TITLE
Installing emsdk overwrites node on PATH

### ISSUE THREAD

**Issue by romatthe** [emscripten-core/emsdk#1142]
This is a small but annoying issue. When installing the emscripten sdk, or whenever you install a new version of the toolchain, you're told to execute `source ./emsdk_env.sh`.

This is rather annoying, as the toolchain comes packaged with its own specific version of node, and this conflicts with all other node tools we're used to using. For example, it put node 4.1.1 or my path while I previously had 7.4.0. I've noticed the same goes for `clang`, but it doesn't bother my that much. Even when relying on a tool like nvm, it's annoying because switching to another version would essentially break emscripten/emsdk?

If emscripten is reliant on a specific version of node, should it not be its own responsibility to make sure its scripts run with its own internally packaged version? This just seems a tad messy.

(**Note:** One possible temporary workaround is to simply avoid `emsdk_env.sh` in this case and just manually add `emsdk/upstream/emscripten` to your path.  That is all that should be needed to get emscripten working on many systems).

**Comment by sbc100** [comment]
> > Is there no workaround for this issue as of today?
> 
> See [#1142 (comment)](https://github.com/emscripten-core/emsdk/issues/1142#issuecomment-1334064978).

IIUC the ideal solution doesn't avoid installing the emsdk version of node, or have emscripten use an unexpected/system version.  The goal is to avoid having emsdk's override system node in the PATH, right?   Avoiding emsdk node, even within emsdk itself, seems like overkill and could cause other issues.

**Comment by kleisauke** [comment]
> IIUC the ideal solution doesn't avoid installing the emsdk version of node

As mentioned in that comment, you cannot activate Emscripten without installing the bundled Node.js provided by emsdk.

> or have emscripten use an unexpected/system version.

This assumes that the globally-installed Node.js version is newer than what emsdk provides. emsdk currently provides Node.js 14.x, which active support ended one year ago. Security updates for this version ends in 5 months. So, I would argue to at least update the bundled Node.js version to the latest LTS version, which would also fix issues like https://github.com/emscripten-core/emscripten/issues/18084.

FWIW, this workaround is currently used in at least libjxl and wasm-vips to avoid compatibility issues with Wasm SIMD, which requires Node.js 16.4.0 or later to match the final SIMD opcodes.
https://github.com/libjxl/libjxl/blob/a58dd9de85f512d6b8870d2ab40caa8423ae0e51/.github/workflows/build_test.yml#L401
https://github.com/kleisauke/wasm-vips/blob/e89ea44d7376c0bbd35cc3ff28d79d97a6340104/Dockerfile#L23-L24

> The goal is to avoid having emsdk's override system node in the PATH, right?

I agree that would be the ideal solution. I only pointed out that comment since he was looking for a workaround.

> Avoiding emsdk node, even within emsdk itself, seems like overkill and could cause other issues.

Just wondering, what do you mean with other issues? With the same assumption as above, at least for Emscripten, using  a globally-installed Node.js should be safe after PR https://github.com/emscripten-core/emscripten/pull/18070. 

Another solution would be that emsdk only installs the bundled Node.js, if a globally-installed Node.js is not available, or if the globally-installed Node.js version is lower than the minimum required Node.js version (or the one provided by emsdk). This would also fix this `PATH` clobbering issue in most cases.

**Comment by sbc100** [comment]
> Just wondering, what do you mean with other issues? With the same assumption as above, at least for Emscripten, using a globally-installed Node.js should be safe after PR [emscripten-core/emscripten#18070](https://github.com/emscripten-core/emscripten/pull/18070).

We only test emcc itself with a specific version of node, so using any other version is vearing into untested teritory.  Its probably fine, but we only test our internal compiler tools against that fixed version.

> Another solution would be that emsdk only installs the bundled Node.js, if a globally-installed Node.js is not available, or if the globally-installed Node.js version is lower than the minimum required Node.js version (or the one provided by emsdk). This would also fix this `PATH` clobbering issue in most cases.

I tried to land that solution in #714 but it was deemed to magical.  We could try to re-litigate that one.  I think its quite a nice solution myself.

**Comment by kleisauke** [comment]
> We only test emcc itself with a specific version of node, so using any other version is vearing into untested teritory. Its probably fine, but we only test our internal compiler tools against that fixed version.

I understand. There's a `test-latest-node` CI job in Emscripten, but that currently lacks some coverage as it only runs two tests.

> I tried to land that solution in #714 but it was deemed to magical. We could try to re-litigate that one. I think its quite a nice solution myself.

Great, I had forgotten about that PR. IIUC, that only affects the `emsdk activate` phase and still downloads and extracts the Node.js tarball (or zipball on Windows)?

If so, perhaps a better solution would be to do this check in the `emsdk install` phase? We should probably only do this if Emscripten is installed (i.e. only for the tags/aliases listed in [`emscripten-releases-tags.json`](https://github.com/emscripten-core/emsdk/blob/main/emscripten-releases-tags.json)). This way, users can still install and activate the bundled Node.js provided by emsdk via:
```console
$ ./emsdk install node-14.18.2-64bit
$ ./emsdk active node-14.18.2-64bit
```

**Comment by sbc100** [comment]
> > We only test emcc itself with a specific version of node, so using any other version is vearing into untested teritory. Its probably fine, but we only test our internal compiler tools against that fixed version.
> 
> I understand. There's a `test-latest-node` CI job in Emscripten, but that currently lacks some coverage as it only runs two tests.
> 
> > I tried to land that solution in #714 but it was deemed to magical. We could try to re-litigate that one. I think its quite a nice solution myself.
> 
> Great, I had forgotten about that PR. IIUC, that only affects the `emsdk activate` phase and still downloads and extracts the Node.js tarball (or zipball on Windows)?

Yes, in fact it still uses the bundled version of node within emscripten (i.e. it sets NODE_JS in the config file).  All it does it avoid adding this version of node to the PATH during `emsdk_env`.   In this sense it is a very minimal solution to this bug (which is about PATH conflicts, and not so much about want emscripten to use the host node).

> 
> If so, perhaps a better solution would be to do this check in the `emsdk install` phase? We should probably only do this if Emscripten is installed (i.e. only for the tags/aliases listed in [`emscripten-releases-tags.json`](https://github.com/emscripten-core/emsdk/blob/main/emscripten-releases-tags.json)). This way, users can still install and activate the bundled Node.js provided by emsdk via:
> 
> ```
> $ ./emsdk install node-14.18.2-64bit
> $ ./emsdk active node-14.18.2-64bit
> ```

That would be an alternative, and slightly more intrusive/risky change yes.   It would solve a slightly different issue which is "I want emcc to use my system version of node".

**Comment by johnhooks** [comment]
> The simplest solution might just be to avoid using emsdk_env.sh completely. Instead you can simply add emsdk/upstream/emscripten to your PATH.

@sbc100 Is there anyway to pin the current solution to the top of this issue? It took me a while to find this amongst the full discussion.

As a JavaScript developer learning Emscripten, this was a very confusing. I am using Emscripten along with JavaScript/TypeScript build tools to bundle an application and I am using a very recent version of Node.js.

**Comment by sbc100** [comment]
(I don't think there is any way to do that, but I will see if I can edit the description of the add a node).

**Comment by sbc100** [comment]
This was fixed in #1189
